### ISSUE TITLE
WS proxy websocket connection fails with "Invalid frame header" intermittently

### ISSUE THREAD

**Issue by raeesaa** [chimurai/http-proxy-middleware#226]
### Expected behavior
  - Client should always be able to connect to websocket connection via node.js proxy server.

### Actual behavior

Client was able to connect websocket connection via node.js proxy server on deployment but connections started intermittently failing with following errors after few hours:

  - Chrome: "WebSocket connection to 'ws://<server_dns>/subscriptions' failed: Invalid frame header"
  - Safari: "WebSocket connection to 'ws://<server_dns>/subscriptions' failed: Compressed bit must be 0 if no negotiated deflate-frame extension"

And I see following logs at server side:

```
13/12/2017 16:01:49[HPM] Client disconnected
13/12/2017 16:02:00[HPM] Upgrading to WebSocket
13/12/2017 16:02:11[HPM] Client disconnected
13/12/2017 16:02:13[HPM] Upgrading to WebSocket
13/12/2017 16:02:13[HPM] Client disconnected
13/12/2017 16:02:15[HPM] Upgrading to WebSocket
13/12/2017 16:02:50[HPM] Client disconnected
13/12/2017 16:03:00[HPM] Upgrading to WebSocket
13/12/2017 16:03:00[HPM] Client disconnected
13/12/2017 16:03:05[HPM] Client disconnected
13/12/2017 16:03:09[HPM] Upgrading to WebSocket
13/12/2017 16:03:10[HPM] Upgrading to WebSocket
13/12/2017 16:03:10[HPM] Client disconnected
13/12/2017 16:03:30[HPM] Upgrading to WebSocket
13/12/2017 16:03:30[HPM] Client disconnected
13/12/2017 16:03:40[HPM] Upgrading to WebSocket
13/12/2017 16:03:40[HPM] Client disconnected
13/12/2017 16:03:59[HPM] Client disconnected
13/12/2017 16:04:00[HPM] Upgrading to WebSocket
13/12/2017 16:04:00[HPM] Client disconnected
13/12/2017 16:04:05[HPM] Upgrading to WebSocket
13/12/2017 16:04:05[HPM] Client disconnected
13/12/2017 16:04:14[HPM] Upgrading to WebSocket
13/12/2017 16:04:20[HPM] Upgrading to WebSocket
13/12/2017 16:05:04[HPM] Client disconnected
13/12/2017 16:05:10[HPM] Client disconnected
13/12/2017 16:05:15[HPM] Upgrading to WebSocket
```

#### Node module versions:

* http-proxy-middleware: 0.17.4
* server: express 4.14.0

#### Server deployment configuration:

* Node.js server and websocket proxy server, it tries to connect to, are both deployed within docker containers.

#### proxy middleware configuration
```javascript

let subscriptionProxyOption = {
    target: <proxy_server_dns>,
    changeOrigin: true,
    ws: true,
    pathRewrite: {
        '/graphql-subscription': ''
    }
}
var subscriptionProxy = proxy('/graphql-subscription', subscriptionProxyOption);
```
#### server mounting
```javascript
var app = express();

app.use(subscriptionProxy);
let server = app.listen(process.env.PORT);
server.on('upgrade', subscriptionProxy.upgrade)
```

**Comment by chimurai** [comment]
Would be great if you can create a minimal setup in which this issue can be reproduced.

**Comment by raeesaa** [comment]
@chimurai I'll setup a test environment over weekend and let you know. Little busy with other stuff now.

**Comment by raeesaa** [comment]
@chimurai This happens only on my office network. Everything works fine on my home internet. Looks like it has something to do with network proxy / firewall.

**Comment by chimurai** [comment]
Can you try using this?

https://github.com/chimurai/http-proxy-middleware/blob/master/recipes/corporate-proxy.md

**Comment by raeesaa** [comment]
Will give it a try. Thank you :)

**Comment by chimurai** [comment]
Hi @raeesaa , did you find a solution?

**Comment by raeesaa** [comment]
I did get chance to try solution provided for corporate proxy but our IT did some changes in proxy rules and websocket started working fine after that.

Thank you for the great library :)

**Comment by moeiscool** [comment]
@raeesaa what changes were made?

**Comment by dcworldwide** [comment]
Anyone resolve this? It's a constant issue for us, we can't connect to our websocket endpoint via http proxy middleware at all since upgrading from 3

**Comment by alanhg** [comment]
I also encountered a similar problem, what is the specific solution?

![image](https://user-images.githubusercontent.com/9245110/205498842-e483aba4-105f-40e8-8abf-6d0e62cfaed0.png)

WebSocket connections report the error. and no response.

![image](https://user-images.githubusercontent.com/9245110/205498869-365cff8b-9cb2-44ec-b3ab-92be45b8c690.png)


I used an agent. but I think it's not an agent error, some request is ok, just ws.
